advanced_functionality/pytorch_extending_our_containers/pytorch_extending_our_containers.ipynb

```diff
@@ -0,0 +1,701 @@
+    "    └── cifar10\n",
+    "        ├── cifar10.py\n",
+    "\n",
+    "Let's discuss each of these in turn:\n",
+    "\n",
+    "* __`Dockerfile`__ describes how to build your Docker container image. More details are provided below.\n",
+    "* __`build_and_push.sh`__ is a script that uses the Dockerfile to build your container images and then pushes it to ECR. We invoke the commands directly later in this notebook, but you can just copy and run the script for your own algorithms.\n",
+    "* __`cifar10`__ is the directory which contains our user code to be invoked.\n",
+    "\n",
+    "In this simple application, we install only one file in the container. You may only need that many, but if you have many supporting routines, you may wish to install more.\n",
+    "\n",
+    "The files that we put in the container are:\n",
+    "\n",
+    "* __`cifar10.py`__ is the program that implements our training algorithm and handles loading our model for inferences."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### The Dockerfile\n",
+    "\n",
+    "The Dockerfile describes the image that we want to build. You can think of it as describing the complete operating system installation of the system that you want to run. A Docker container running is quite a bit lighter than a full operating system, however, because it takes advantage of Linux on the host machine for the basic operations. \n",
+    "\n",
+    "We start from the SageMaker PyTorch image as the base. The base image is an ECR image, so it will have the following pattern.\n",
+    "* {account}.dkr.ecr.{region}.amazonaws.com/sagemaker-{framework}:{framework_version}-{processor_type}-{python_version}\n",
+    "\n",
+    "Here is an explanation of each field.\n",
+    "1. account - AWS account ID the ECR image belongs to. Our public deep learning framework images are all under the 520713654638 account.\n",
+    "2. region - The region the ECR image belongs to. [Available regions](https://aws.amazon.com/about-aws/global-infrastructure/regional-product-services/).\n",
+    "3. framework - The deep learning framework.\n",
+    "4. framework_version - The version of the deep learning framework.\n",
+    "5. processor_type - CPU or GPU.\n",
+    "6. python_version - The supported version of Python.\n",
+    "\n",
+    "So the SageMaker PyTorch ECR image would be:\n",
+    "520713654638.dkr.ecr.us-west-2.amazonaws.com/sagemaker-pytorch:0.4.0-cpu-py3\n",
+    "\n",
+    "Information on supported frameworks and versions can be found in this [README](https://github.com/aws/sagemaker-python-sdk).\n",
+    "\n",
+    "Next, we add the code that implements our specific algorithm to the container and set up the right environment for it to run under.\n",
+    "\n",
+    "**DISCLAIMER: As of now, the support for the two environment variables below are only supported for the SageMaker Chainer (4.1.0+) and PyTorch (0.4.0+) containers.**\n",
```

Comment:
The variables would work if you add RUN pip install sagemaker-containers, right?

Reply:
For our framework containers, not necessarily because we changed the name of methods here.
https://github.com/aws/sagemaker-containers/blob/master/src/sagemaker_containers/_modules.py#L274
https://github.com/aws/sagemaker-pytorch-container/blob/master/src/sagemaker_pytorch_container/training.py#L50

Reply:
I think we should either add some more markdown about a workaround if users wanted to do this with the TensorFlow and MXNet containers, or I think we should remove the mention of those frameworks at the top of the notebook and make it clear that this is functionality that only applies to PyTorch and Chainer right away.

Reply:
Yes, this notebook should be updated when the TF and MXNet containers can handle this change.